### ISSUE TITLE
Change outputChannelCount dynamically after instantiation (and independently from input channelCount)

### ISSUE THREAD

**Issue by JohnWeisz** [WebAudio/web-audio-api#1780]
## Describe the feature

Allow changing [outputChannelCount](https://www.w3.org/TR/webaudio/#dom-audioworkletnodeoptions-outputchannelcount) dynamically, after and separately from instatiation.

The following example briefly describes a node which always takes a single stereo input, has a single stereo output (by default), and can switch between stereo output, quad output, and auto output on demand, dynamically, at any time:

```
class CustomChannelMixerNode extends AudioWorkletNode {
    constructor() {
        super("custom-channel-mixer-processor", {
            numberOfInputs: 1,
            numberOfOutputs: 1,
            channelCountMode: "explicit",
            channelCount: 2,
            outputChannelCount: [2]
        });
    }

    useStereoOutput() {
        if (this.outputChannelCount && this.outputChannelCount[0] === 2) {
            console.warn("already using stereo output.");
            return;
        }

        this.outputChannelCount = [2];
    }

    useQuadOutput() {
        if (this.outputChannelCount && this.outputChannelCount[0] === 4) {
            console.warn("already using quad output.");
            return;
        }

        this.outputChannelCount = [4];
    }

    useAutoOutput() {
        if (!this.outputChannelCount) {
            console.warn("already using auto output.");
            return;
        }

        this.outputChannelCount = null;
    }

    // ...
}
```

## Is there a prototype?

Can be done in a limited fashion using a wrapper object (and augmenting the `AudioNode.connect` method to accept the wrapper object), as currently a reinstatiation of the AudioWorkletNode is required to change output channel count.

### General behavior

Assigning to `this.outputChannelCount` will result in the exact same behavior as if the AudioWorkletNode was initialized with the specified output channel configuration in the first place, beginning with the next `process` call on the corresponding AudioWorkletProcessor instance.

### Immutability

Assigning to `this.outputChannelCount` is consistent with assigning to `this.channelCount`, although it is up to question how exactly assigning to a single specific output is handled, i.e. what does the following do:

```
this.outputChannelCount[0] = 4;
```

My vote here is to freeze `this.outputChannelCount`, and always require a full array assignment.

### Clearing an explicitly defined value

In a case where the developer wants to clear any values set previously, and instead use an automatically determined outputChannelCount, `this.outputChannelCount` can be assigned `null`. _Personally, I feel this part of the proposal needs some more discussion._

**Comment by JohnWeisz** [comment]
## Potential use cases and optimization possibilities

 - Most importantly: external plugin systems where plugins can adjust their input/output count dynamically and independently from each other, and no further assumptions can be made about how the plugin interprets channel layouts internally
 - Reverb and convolver effects where, depending on a dynamically changeable algorithm, the output channel count can change, while the number of inputs remain the same
 - Where discrete channel mixing is used and simply picking `max(desiredInputChannelCount, desiredOutputChannelCount)` as `this.channelCount` would suffice, having separately adjustable input/output channel count can facilitate reducing the number of `Float32Array` objects allocated, thus reducing GC and improving performance

**Comment by karlt** [comment]
Comparing the API proposed here with the mutation of outputs suggested in https://github.com/WebAudio/web-audio-api/issues/1515#issuecomment-406173462 for similar purposes:

Mutation of `outputs` allows `process()` to choose and output channel count based on inputs.  With the API proposed here, such a change would not be possible (at least not without a delay of one rendering quantum).

**Comment by hoch** [comment]
From the abstraction point of view, using a property rather than internal mutation makes more sense to me. It is possible that the actual user (e.g. plug-in user) might not have access to the code of AudioWorkletProcessor (e.g. developed by third-party). Hiding the channel mutation logic inside of the processor will confuse users with the processor's mysterious behavior.

**Comment by padenot** [comment]
Both solution have merit: the algorithm can certainly do some optimizations based on input channel count and internal state, to output a specific number of channels.

However it can be useful to set it from outside, as a control parameter to the algorithm.

Both make sense, maybe both are needed in fact.

**Comment by hoch** [comment]
To track it properly, I am putting this to the V1 bucket.

**Comment by hoch** [comment]
If both ideas make make sense, then we have a conflict on our hands: what if both happen at the roughly same time? Each operation happens on a different thread. The property change happens on the main thread and will be posted to the rendering thread later, and the internal array mutation happens on the rendering thread. In short, this will be a flaky situation, which we don't want to have more of it.

After more thoughts, now I am actually in favor of the internal array mutation. We already have some native AudioNodes that have opaque channel behavior. It is not ideal, but justifiable.

**Comment by hoch** [comment]
An example:

```js
const getNewOutputChannels = (channelCount) => {
  const output = [];
  for (let i = 0; i < channelCount; ++i)
    output.push(new Float32Array(128));
  return output;
};

...

/* AudioWorkletProcessor */
process(inputs, outputs) {
  const input = inputs[0];
  const output = getNewOutputChannels(4);
  for (let channel = 0; channel < output.length; ++channel) {
    // do something with |output|.
  }
  outputs[0] = output;
  return true;
}
```

If this is what we want, the spec change can be minimal actually. We simply can say that the output array will be mutable, and the output channel count will be changed accordingly. FWIW, this will trigger the downstream channel count change as well.

**Comment by hoch** [comment]
cc @rtoy

**Comment by padenot** [comment]
I agree with changing the output channel count by mutating the output array.

**Comment by hoch** [comment]
Thanks for your feedback, @JohnWeisz!

The WG has decided to push this issue to v.next. Until then we'll be collecting more field data from the developers to figure out the right design for this feature. It could be either using a node side property,  dynamic mutation of output array or something else.

**Comment by rtoy** [comment]
Per https://github.com/WebAudio/web-audio-api/issues/1780#issuecomment-459434942 really make it v.next.
